### PR TITLE
Better error for str.cat with listlike of wrong dtype.

### LINKED ISSUE #22722: Improve TypeError message for str.cat
Currently, 

```
s = pd.Series(['a', 'b', 'c'])
s.str.cat([1, 2, 3])
```
yields
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Axel Obermeier\eclipse-workspace\pddev\pandas\core\strings.py", line 2222, in cat
    res = str_cat(data, others=others, sep=sep, na_rep=na_rep)
  File "C:\Users\Axel Obermeier\eclipse-workspace\pddev\pandas\core\strings.py", line 111, in str_cat
    cats = [sep.join(tup) for tup in tuples]
  File "C:\Users\Axel Obermeier\eclipse-workspace\pddev\pandas\core\strings.py", line 111, in <listcomp>
    cats = [sep.join(tup) for tup in tuples]
TypeError: sequence item 1: expected str instance, int found
```

IMO, this should be improved to have a better error message, and shallower stack trace.

### PR DESCRIPTION
**Pull request by h-vetinari**
- [x] closes #22722
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This had been blocked on #23167.

### CODE REVIEW

File: pandas/core/strings.py
Comment (gfyoung) [inline pandas/core/strings.py]: No exclamation points

File: pandas/core/strings.py
Comment (gfyoung) [inline pandas/core/strings.py]: This logic looks somewhat complicated.  Add a comment explaining why we need this type of check.

File: pandas/tests/test_strings.py
Comment (gfyoung) [inline pandas/tests/test_strings.py]: Reference issue number as a comment above

File: doc/source/whatsnew/v0.25.0.rst
Comment (gfyoung) [inline doc/source/whatsnew/v0.25.0.rst]: * Just say "Improved error message when passing..."

* Backticks around `Series`

File: pandas/core/strings.py
Comment (simonjayhawkins) [reply pandas/core/strings.py]: even though it's now documented, i think i would still prefer to see this split somehow.

File: pandas/tests/test_strings.py
Comment (simonjayhawkins) [inline pandas/tests/test_strings.py]:
```suggestion
    def test_str_cat_wrong_dtype_raises(self, box):
```

File: pandas/tests/test_strings.py
Comment (simonjayhawkins) [inline pandas/tests/test_strings.py]: is the .* needed?

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
@h-vetinari my main concern is the complexity of the condition.

File: pandas/tests/test_strings.py
Comment (h-vetinari) [reply pandas/tests/test_strings.py]: I like to end regexes for long warning messages with `.*` to make it abundantly clear that the message actually continues.

File: pandas/tests/test_strings.py
Comment (h-vetinari) [reply pandas/tests/test_strings.py]: Fair

File: pandas/core/strings.py
Comment (h-vetinari) [reply pandas/core/strings.py]: Will try.

File: pandas/tests/test_strings.py
Comment (simonjayhawkins) [inline pandas/tests/test_strings.py]: do these cover the different conditions for raising?

Comment (h-vetinari) [reply pandas/tests/test_strings.py]: IMO yes. It would be possible to add more dtypes, but I don't consider this a great improvement.

Happy to add if you don't mind combinatorially more tests.

Comment (simonjayhawkins) [reply pandas/tests/test_strings.py]: so if i modify the condition, we have tests that fail?

Comment (h-vetinari) [reply pandas/tests/test_strings.py]: I don't understand what you mean here. Adding different containers (as long as they are legal inputs to `.str.cat`) is possible, but I've already added the most import ones.

It would be possible add a further parametrization over the dtype and not just test that integer Series fail (as currently), but also floats, complex, datetime etc. This would work without failure.

You'd have to be  more concrete with "modify the condition" for me to answer more clearly. This test serves as a relatively minimal example from #22722. If you want more dtypes, I'm happy to add them. Anything else you'll need to explain to me in more detail.

Comment (simonjayhawkins) [reply pandas/tests/test_strings.py]: not raising the message on an invalid case would just result in the old message be raised. rasing on valid cases is a regression.

would it not be simpler to catch the message and reraise?

Comment (h-vetinari) [reply pandas/tests/test_strings.py]: > would it not be simpler to catch the message and reraise?

In principle, yes. But:
* there are 3 separate calls to `cat_core` that would need to be wrapped
* performance: I don't want to do all the gymnastics of aligning, `ensure_object`, calculating and unioning na_masks and cat_core itself if we already (could) know beforehand that things will fail.

> not raising the message on an invalid case would just result in the old message be raised. rasing on valid cases is a regression.

It is possible to circumvent the check I'm adding. `pd.Series(['a', 1, 'b'])` will fail with the old error, but we wouldn't be able to tell from the dtype beforehand. Raising on valid cases cannot happen I believe - all valid concatenations must be either strings or missing values (that's why the all-NA float case is in there, which can happen for example due to mis-alignment; as opposed to e.g. complex, which could have all-NA but shouldn't [be expected to] work with strings)

Comment (simonjayhawkins) [reply pandas/tests/test_strings.py]: @h-vetinari thanks for the explanation. i'm not convinced that for the sake of a better error message we should be adding this complexity, especially if the old message is still raised in some cases.

@jreback

Comment (h-vetinari) [reply pandas/tests/test_strings.py]: > i'm not convinced that for the sake of a better error message we should be adding this complexity, especially if the old message is still raised in some cases.

I think looking at #22722, this is a clear win - it covers 99% (guesstimate) of real-world cases where this error happens. If someone actually has mixed strings / integers in their Series, it just won't fail as gracefully. But that doesn't imply (to me) that we shouldn't aim to do better in the 99%.

Comment (h-vetinari) [reply pandas/tests/test_strings.py]: Side note: if your only worry is an inconsistent error message, I could have the 99%-fail-early-check *and* catch-and-reraise afterwards (with the same error message).

Comment (simonjayhawkins) [reply pandas/tests/test_strings.py]: best to wait for feedback from other maintainers before you make any changes on my account.

Comment (h-vetinari) [reply pandas/tests/test_strings.py]: Too late. ;-)

File: pandas/core/strings.py
Comment (simonjayhawkins) [inline pandas/core/strings.py]: if we reach here, we already have a user-facing TypeError? and we are just replacing the message.

are any other TypeErrors expected, i'd be inclined to not check the message check.

if it is necessary, maybe have a `_is_not_string_exception` in pandas\core\dtypes\common.py similar to `_is_unorderable_exception` rather than have the condition here.

Comment (h-vetinari) [reply pandas/core/strings.py]: > are any other TypeErrors expected, i'd be inclined to not check the message check.

Fair enough, that's a not gonna happen.

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use ``:class:`Series` ``

File: pandas/core/strings.py
Comment (jreback) [inline pandas/core/strings.py]: can you limit the try/except to the relevant code

File: pandas/core/strings.py
Comment (jreback) [inline pandas/core/strings.py]: this is way complicated, what exactly are you trying to do here

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst
Comment (h-vetinari) [reply doc/source/whatsnew/v0.25.0.rst]: No problem

File: pandas/core/strings.py
Comment (h-vetinari) [reply pandas/core/strings.py]: I originally had an inline condition within `any`, but @simonjayhawkins found this too complex, so I broke out that condition into a function.

Basically, I want to fail early for any Series that will necessarily fail concatenation (based on dtype). Object must obviously be allowed, but also all-NA float (which can happen if two Series completely misalign), plus needs handling of categorical.

File: pandas/core/strings.py
Comment (h-vetinari) [reply pandas/core/strings.py]: I am; `cat_core` is called in all three branches, and may raise in any one of them.

**Review by h-vetinari — COMMENTED** [review]
Responses

File: pandas/core/strings.py
Comment (jreback) [reply pandas/core/strings.py]: do you not already have the inferred types at this point? and if you don't, why not just infer them, then this condition becomes easier

File: pandas/core/strings.py
Comment (jreback) [reply pandas/core/strings.py]: so rather than having a giant try/except, either add this check in the cat_core, or write a function which *calls* cat_core and catches and formats a nicer error

maybe try to combine the _legal_dtype check with this, IOW. you can just try to cat them then catch an error, at which point you can then infer and give a nice message.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/strings.py
Comment (simonjayhawkins) [reply pandas/core/strings.py]: > I originally had an inline condition within `any`, but @simonjayhawkins found this too complex, so I broke out that condition into a function.

the any is applied to a generator expesssion with a for loop and then raising. so i'm not sure the use of any here is a benefit. could you not just use a for loop and do away with the separate function?

**Review by h-vetinari — COMMENTED** [review]
@jreback @simonjayhawkins 

I moved the try/except + raise into a separate wrapper, as @jreback suggested. This allows to remove the extra check block I had introduced.

However, removing that extra check has the edge case that wrongly dtyped data that is completely misaligned will slip through (i.e. not raise):
```
>>> s = pd.Series(['a', 'b', 'c'])
>>> t = pd.Series([0, 1, 2], index=[10, 11, 12])
>>> s.str.cat(t, join='left')
0    NaN
1    NaN
2    NaN
dtype: object
```

It is a downside, but I would be fine with it, not least considering the much less complicated code now.

File: pandas/core/strings.py
Comment (jreback) [inline pandas/core/strings.py]: can you type the args  & return value & add a Parameters / Returns section

File: pandas/core/strings.py
Comment (jreback) [inline pandas/core/strings.py]: move the comment to the except

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use ``:class:`Series` ``

**Review by jreback — CHANGES_REQUESTED** [review]
looks much more reasonable. typing & doc-string comments, ping on green.

File: pandas/core/strings.py
Comment (simonjayhawkins) [inline pandas/core/strings.py]: Simple is better than complex.

what's wrong with:

```python
        for column in list_of_columns:
            dtype = lib.infer_dtype(column, skipna=True)
            if dtype not in ('string', 'empty'):
                raise TypeError('Concatenation requires list-likes containing only '
                                'strings (or missing values). Offending values found '
                                'in column {}'.format(dtype)) from None
        raise
```

File: pandas/core/strings.py
Comment (h-vetinari) [reply pandas/core/strings.py]: What's your expectation for typing the args? Just `List`? It would strictly speaking be `List[np.array]`, but AFAICT, mypy resp. the typing module doesn't yet support numpy stubs natively.

**Review by h-vetinari — COMMENTED** [review]
Typing comments

**Comment by simonjayhawkins** [comment]
@h-vetinari if you merge master the checks ci failure should be fixed. I think i've seen the Windows py37_np141 failure before. probably a flaky test. likely pass next time around.

File: pandas/core/strings.py
Comment (jreback) [reply pandas/core/strings.py]: yes that would be fine

**Review by jreback — CHANGES_REQUESTED** [review]
typing request, otherwise lgtm. ping on green.

File: pandas/core/strings.py
Comment (simonjayhawkins) [inline pandas/core/strings.py]: is it worth having a `raise` outside the for loop to ensure we don't slip through, or is that not going to happen?

**Review by simonjayhawkins — APPROVED** [review]
much more comfortable without the pre-check to fail early and simpler code.

Thanks @h-vetinari

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @h-vetinari

**Pull request merged**
